readthedocs/search/views.py

```diff
@@ -43,84 +45,117 @@ def elastic_search(request, project_slug=None):
     :param project_slug: Sent when the view is a project search
     """
 
-    request_type = None
-    if project_slug:
-        queryset = Project.objects.protected(request.user)
-        project_obj = get_object_or_404(queryset, slug=project_slug)
-        request_type = request.GET.get('type', 'file')
-
-    user_input = UserInput(
-        query=request.GET.get('q'),
-        type=request_type or request.GET.get('type', 'project'),
-        project=project_slug or request.GET.get('project'),
-        version=request.GET.get('version', LATEST),
-        language=request.GET.get('language'),
-        role_name=request.GET.get('role_name'),
-        index=request.GET.get('index'),
-    )
-    search_facets = collections.defaultdict(
-        lambda: ProjectSearch,
-        {
-            'project': ProjectSearch,
-            'file': PageSearch,
-        }
-    )
-
-    results = []
-    facets = {}
+    http_method_names = ['get']
 
-    if user_input.query:
-        filters = {}
+    def _get_project(self, project_slug):
+        queryset = Project.objects.protected(self.request.user)
+        project = get_object_or_404(queryset, slug=project_slug)
+        return project
 
-        for avail_facet in ALL_FACETS:
-            value = getattr(user_input, avail_facet, None)
-            if value:
-                filters[avail_facet] = value
-
-        search = search_facets[user_input.type](
-            query=user_input.query,
-            filters=filters,
-            user=request.user,
+    def _get_project_data(self, project, version_slug):
+        version_doctype = (
+            project.versions
+            .values_list('documentation_type', flat=True)
+            .get(slug=version_slug)
         )
-        results = search[:50].execute()
-        facets = results.facets
+        project_data = {
+            project.slug: (
```

Comment:
Yea, I think an explicit data structure would be nice here, as well. Much clearer the intent.